Updated lesson plan for expressions, types and values

zainab-ali opened the pull request:
I've included some questions which were useful when working through this.  

It might also be worth adding a list of topics to avoid, notably:
 - parentheses in arithmetic expressions (they confuse these with parentheses in method calls)
 - references to the terms compile time, run time and the compiler.  "Before evaluation" and "during evaluation" should be enough to get their mental model started.
 - Too much detail on the model of execution.  @yilinwei may have other opinions on this (?)

Let me know what you think

zainab-ali commented on level1/expressions-types-values.md:
```diff
@@ -68,7 +68,7 @@ A type is a property of an expression, not of a value. If you have used a "dynam
 > What is the type of `1/0`?
 
-*They should be able to reason about this (expressions have types) but could also check it empirically by running the program.*
+*They should be able to reason about this (expressions have types) but could also check it empirically by running the program.  Discuss this, but don't try to correct their answer.  We haven't yet explained methods*
```
Introducing types without introducing methods was pretty confusing.  Perhaps we should leave the explanation at categorization of values, and go on to error prevention when we introduce methods?

noelwelsh replied:
Interesting. I have mostly worked with students who have some prior programming experience so perhaps they have been more willing to use a prior mental model to explore this question. I'm slightly inclined to keep this. I think it is a good example of the difference between types and tags for those students who have the background to understand it. Perhaps we should have a note to instructors explaining that not all students will be able to make progress here?

zainab-ali replied:
I agree with keeping this for those with prior programming experience.  

We had a group of two students with no experience, one of whom didn't have an IDE to clearly indicate compile time vs runtime errors.  It's not surprising they found this more challenging than a more mixed group.

zainab-ali commented on level1/expressions-types-values.md:
```diff
@@ -100,38 +100,201 @@ Here are three general problem solving strategies:
 ## Objects and Method Calls
 
-See Creative Scala Chapter 2
+The operations on an object are called it's methods.
```
I'm not sure if it's worth talking about "objects" at all.  What do you think about saying that operations are defined by the type of expression, and leaving out the term "object" until we introduce the idea of data?

noelwelsh replied:
We need to introduce objects fairly early, because they have to wrap everything in an object to get it compile. We don't have to introduce the term now, though. I'm neutral here.

zainab-ali replied:
I would favour leaving it out and introducing it early in the next chapter.  This lesson seems jargon-heavy to those with no prior programming experience.

zainab-ali commented on level1/expressions-types-values.md:
```diff
@@ -100,38 +100,212 @@ Here are three general problem solving strategies:
-3. Write code to draw three circles in a row (that is, one circle beside another circle beside another circle). Make the circles different colours.
-
-
-4. Write a program that draws four circles in a square pattern. Make each circle a different color.
```
I'm assuming they haven't seen doodle at this point, so am removing the references to it.

noelwelsh replied:
They should have seen the example program in the Creative Scala template project. The main aim with this exercise is to do something a bit more fun than arithmetic (and string munging). The secondary aim is to stretch them just a little bit, encouraging them to make inferences from the existing code in the template project. Not all students will be able to do this, however.

zainab-ali commented on level1/expressions-types-values.md:
```diff
@@ -100,25 +100,205 @@ Here are three general problem solving strategies:
+
+Which parts use method call syntax?  Which use infix notation?
+
+Can you write this completely in method call syntax?
```
Maybe we should just call it "method call notation" throughout, and avoid the term "syntax"?

noelwelsh replied:
We do need to introduce the term at some point, as students will encounter it in other contexts. I think Creative Scala has more detail on this, but I'm not 100% certain.

zainab-ali replied:
Creative Scala doesn't ever define the term *syntax*, and starts using it when describing method call syntax.  I do think it's important to explore notation at this point, especially for those with no prior experience, as they cannot distinguish between the textual representation of an expression and the expression itself.  Using different notation for the same expression nudges them towards this.

Perhaps syntax is something we should give a concrete definition to?

noelwelsh replied:
Yes, that sounds sensible to me. I've opened an issue on Creative Scala to define "syntax" in there.

zainab-ali commented:
@noelwelsh thanks for the review.  I've added the doodle questions back in (I agree that they should be able to see and run the code, and that we should encourage them to experiment with it).  
I've removed all references to objects and attempted to clarify a few of the points.  Let me know what you think.

noelwelsh commented:
Thanks!
